Added Cloud Configuration Hierarchy (#8031)
magento/devdocs: src/cloud/live/sens-data-over.md

## Consistent configuration

> The database contains default configurations for your Magento store. When you update configurations in the {{site.data.var.ece}} environments using the Magento Admin panel or the Magento CLI command  `bin/magento config:set`, your configuration changes apply to the `app/etc/config.php` file. You can then use this file to manage and synchronize the Magento application configuration for your local environment and across each Cloud environment.

Reviewer: ```suggestion
The database contains default configurations for your Magento store. When you update configurations in the {{site.data.var.ece}} environments using the Magento Admin panel or the Magento CLI `bin/magento config:set` command, your configuration changes apply to the `app/etc/config.php` file. You can then use this file to manage and synchronize the Magento application configuration for your local environment and across each Cloud environment.
```

Author: I like that - we never set values that way so I overlooked it. Thanks.

> The database contains default configurations for your Magento store. When you update configurations in the {{site.data.var.ece}} environments using the Magento Admin panel or the Magento CLI command  `bin/magento config:set`, your configuration changes apply to the `app/etc/config.php` file. You can then use this file to manage and synchronize the Magento application configuration for your local environment and across each Cloud environment.

Reviewer 2: ```
When you update configurations in the {{site.data.var.ece}} environments using the Magento Admin panel or the Magento CLI command  `bin/magento config:set`, your configuration changes apply to the `app/etc/config.php` file
```
If customer changes configurations in Magento Admin panel - these changes do not affect/apply in `app/etc/config.php`. The same for  `bin/magento config:set` all configurations are saved to DB by default. Only if you use option `--lock-config` then values of configuration will be saved to `app/etc/config.php`.

### Sensitive data

Sensitive values are _not_ stored in the `app/etc/config.php` file. Any sensitive configurations export to the `app/etc/env.php` file when you use the `bin/magento app:config:dump` Magento CLI command. For Cloud projects, we recommend using {{ site.data.var.ece }} environment variables to store sensitive values. For an example, see [Add Magento authentication keys][auth].

> You can also set sensitive values using the Magento CLI command : `bin/magento config:sensitive:set`, see [Sensitive or system-specific settings] in the _Configuration Guide_.

Reviewer 3: Remove extra space.

```suggestion
You can also set sensitive values using the Magento CLI command: `bin/magento config:sensitive:set`, see [Sensitive or system-specific settings] in the _Configuration Guide_.
```

> Sensitive values are _not_ stored in the `app/etc/config.php` file. Any sensitive configurations export to the `app/etc/config.php` file during the `config:dump` process. We recommend using environment variables to store sensitive values. For an example, see [Add Magento authentication keys][auth].

Reviewer 2: sensitive configurations export to the `app/etc/env.php` file, not to `app/etc/config.php`

Reviewer 2: Also, there are some other info which is not clear for me. I cannot add commend for that line, but it for lines 34-36.
```
-  Configured values for settings entered through the Magento Admin store configuration
```
config.php file won't have data which was added through Magento Admin panel until running `dump` command. So, that sentence more false than true.
```
- Extension list
```
I do not understand what is meant under this. List of third-party modules? Then yes, it's true as config.php is the only source about modules states (enabled/disabled)
```
- Scopes value for static content deployment (the default SCD strategy is [quick][])
```
why strategy is mentioned here?

P.S. I apologize that I add comments to not touched in this PR info. Probably we may create new PR for this

Author: @NadiyaS - This is outside the remit of this PR. I'd be happy if we can deal with what's in scope for this PR and then open a New PR to overhaul this page - this statement is actually confusing as it's referring to the "ece-tools config:dump" action not the "bin/magento app:config:dump"

If you are okay to look at this page as a whole - I will create an issue right now and we can work on it.

Reviewer 2: @BarnyShergold sure, let's update it in new PR. I will check whole page

> Sensitive values are _not_ stored in the `app/etc/config.php` file. Any sensitive configurations export to the `app/etc/config.php` file during the `config:dump` process. We recommend using environment variables to store sensitive values. For an example, see [Add Magento authentication keys][auth].

Reviewer 3: ```suggestion
Sensitive values are _not_ stored in the `app/etc/config.php` file. Any sensitive configurations export to the `app/etc/env.php` file during the `config:dump` process. We recommend using environment variables to store sensitive values. For an example, see [Add Magento authentication keys][auth].
```

> You can also set sensitive values using the Magento CLI command `bin/magento config:sensitive:set`. See [Sensitive or system-specific settings].

Reviewer: ```suggestion
Also, you can use the Magento CLI `bin/magento config:sensitive:set` command to set sensitive values. See [Sensitive or system-specific settings] in the _Configuration Guide_.
```

Author: @hguthrie  - I don't like the rewording for the first part (for me the 'noun' is "Magento CLI command" so it gets split) - it doesn't scan well for me. But happy to add the "..in the _Configuration Guide_". What do you say?

Reviewer: Ending a sentence with a formatted command can be confusing because of the period. Some people will add the period in the command and then wonder why it doesn't work.

If one must end a sentence with a command as such, I suggest terminating with a colon and then the command...no period.

```
You can set sensitive values using the Magento CLI command: `bin/magento config:sensitive:set`
```

Author: That's a good point about the period. How about -
You can also set sensitive values using the Magento CLI command : `bin/magento config:sensitive:set`,  see [Sensitive or system-specific settings].

> 1. If an environment variable does not exist, use the configuration from a `MAGENTO_CLOUD_RELATIONSHIPS` name-value pair in the [`.magento.app.yaml` file][app-yaml]. Ignore the default configuration.

Reviewer 2: Why `MAGENTO_CLOUD_RELATIONSHIPS` is mentioned here? This variable contains only connection data to services not system configurations.

Author: See above.

1. If an environment variable does not exist and `MAGENTO_CLOUD_RELATIONSHIPS` does not contain a name-value pair, remove all customized configuration and use the values from the default configuration.

If the same setting is configured in multiple places, Magento relies on the following configuration hierarchy to determine which value to apply to the environment :

|Priority|Configuration<br/>Method|Description|

> |---|---|---|

Reviewer: This table seems to have an unnatural order. My spidey-sense wants to see priority listed in some sequential order. Is there a method you used that I overlooked or can this be reordered?

Reviewer 3: The strange order seems to be something that happened in my VS code instance.  I'm not sure what happened. Should be in sequential order.

|3 | `app/etc/env.php` | Environment-specific configuration values stored here are added by using the Magento `app:config:dump` command. Set the system-specific and sensitive values using environment variables or the Magento CLI. See [Sensitive data](#sensitive-data). The `env.php` file is **not** included in source control.|
|4 | `app/etc/config.php` | Values stored here are added by using the Magento `app:config:dump` command. Shared configuration values are added to `config.php`. Set shared configuration from the Magento Admin or using the Magento CLI. See [Consistent configuration](#consistent-configuration). The `config.php` file is included in source control.|
|5 | Database | Values stored here are added by setting configurations in the Magento Admin. Note that configurations set using any of the preceding methods are locked (greyed out) and cannot be edited from the Magento Admin.|
|6 | `config.xml` | Many configurations have default values set in a the `config.xml` file for a module. If Magento cannot find any value set by any of the preceding methods, it falls back to the default value, if set.|

> 

Reviewer 2: I would also mention .magento.env.yaml as data from some configurations (e.g. service configurations) will override data in env.php

Author: Agreed but that is separate from this. This is a table for Magento application configuration hierarchy. This page is about "Configuration management for store settings" not service configurations. I had a long conversation with @meker12 last night and .magento.env.yaml is a different type of configuration (service & platform).

Reviewer 2: @BarnyShergold agree. This is about system configurations only. Probably I was confused by `MAGENTO_CLOUD_RELATIONSHIPS` mentioned above

Author: Yes - I think this whole page needs a rework, especially after your comments earlier. I'm going to raise an issue so we look at it. 👍

1. Add, commit, and push code changes to update a remote environment.

## Configuration Hierarchy

> With Cloud Commerce there are several ways to set admin configurations and it is import to understand the order of priority, if a setting is added in two or more ways, the highest priority method takes precedence.

Reviewer 3: ```suggestion
On {{ site.data.var.ece }} projects, you can configure Magento Admin settings in several ways. If the same setting is configured in multiple places, Magento relies on the following configuration hierarchy to determine which value to apply to the environment. 
```

Author: I like this much better 👍

The current hierarchy is as follows, the highest priority is top of the list.

-  Cloud Admin UI
-  .magento.app.yaml
-  env.php
-  config.php

> -  Database

Reviewer 3: @BarnyShergold  I'm not sure about this section.  I think some of this information is conveyed in the [Configuration logic flow](https://devdocs.magento.com/cloud/live/sens-data-over.html#configuration-logic-flow) section. That seems like the place to add any new information. 

So for Cloud -- We recommend configuring through Magento Admin UI, .magento.app.yaml, and using environment variables.   We don't recommend manually editing env.php, config.php, or the database.    I know that there are ways to edit the files manually, and I understand the need to say how the values in different locations are applied.  I'm thinking that enumerating each location in a table with the location, followed by description of how the values are added in each one. 
 Not exactly sure.  What do you think?

[app-yaml]: {{ site.baseurl }}/cloud/project/magento-app.html
[commerce-dump]: {{ site.baseurl }}/guides/v2.3/reference/cli/magento-commerce.html#appconfigdump
[env-yaml]: {{ site.baseurl }}/cloud/project/magento-env-yaml.html
[environment configuration variables]: {{ site.baseurl }}/cloud/project/projects.html#environment-configuration-variables
[export]: {{ site.baseurl }}/config-guide/cli/config-cli-subcommands-config-mgmt-export.html
[Optimize deployment]: {{ site.baseurl }}/cloud/deploy/optimize-cloud-deployment.html
[Pipeline deployment]: {{ site.baseurl }}/guides/v2.3/config-guide/deployment/pipeline/technical-details.html
[quick]: {{ site.baseurl }}/cloud/env/variables-build.html#scd_strategy
[scd]: {{ site.baseurl }}/cloud/deploy/static-content-deployment.html

> [Sensitive or system-specific settings]: {{ site.baseurl }}/guides/v2.4/config-guide/cli/config-cli-subcommands-config-mgmt-set.html

Reviewer: This seems to be a better link: https://devdocs.magento.com/guides/v2.4/config-guide/prod/config-reference-sens.html

```suggestion
[Sensitive or system-specific settings]: {{ site.baseurl }}/guides/v2.4/config-guide/prod/config-reference-sens.html
```

Author: @meker12 added this - but I concur it's the one you have said - adding @meker12 so they know we have agreed to change it.
